### PR TITLE
internal/docs: add pubsub design doc

### PR DESCRIPTION
**Pull request by ijt**
This is the design to address issue #312.

### CODE REVIEW

**Comment by vangent** [comment]
FYI: you can see the rendered .md [here](https://github.com/ijt/go-cloud/blob/pubsubdesign/internal/docs/pubsub/design.md).

File: internal/docs/pubsub/design.md [hunk 1]
```diff
@@ -0,0 +1,915 @@
...
+        case io.EOF:
+            log.Printf("Got ctrl-C. Exiting.")
+            break
+        case nil:
+        default:
+            return err
+        }
+
+        sem <- token{}
+        go func(msg *pubsub.Message) {
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Done.

File: internal/docs/pubsub/design.md [hunk 2]
```diff
@@ -0,0 +1,915 @@
...
+    const poolSize = 10
+    // Use a buffered channel as a semaphore.
+    sem := make(chan token, poolSize)
+    for {
+        msg, err := s.Receive(ctx)
+        switch err {
+        // s.Close() causes io.EOF to be returned from s.Receive().
+        case io.EOF:
+            log.Printf("Got ctrl-C. Exiting.")
+            break
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Done.

File: internal/docs/pubsub/design.md [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+    }
+}
+```
+
+### Driver implementer’s perspective
+Adding support for a new pubsub system involves the following steps, continuing with the "acme" example: 
+
+1. Add a new package called `acmepubsub`.
+2. Add private `topic` and `subscription` types to `acmepubsub` implementing the corresponding interfaces in the `github.com/go-cloud/pubsub/driver` package.
+3. Add a func called `acmepubsub.OpenTopic` that creates an `acmepubsub.topic` and returns a concrete `pubsub.Topic` object make from it.
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Done.

File: internal/docs/pubsub/design.md [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+    "context"
+
+    "github.com/google/go-cloud/pubsub"
+)
+
+// TopicOptions contains configuration for Topics.
+type TopicOptions struct {
+    // SendWait tells the max duration to wait before sending the next batch of
+    // messages to the server.
+    SendWait time.Duration
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Done.

File: internal/docs/pubsub/design.md [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+    // of acknowledgements back to the server.
+    AckWait     time.Duration
+
+    // AckBatchSize is the maximum number of acks that should be sent to
+    // the server in a batch.
+    AckBatchSize int
+
+    // AckDeadline tells how long the server should wait before assuming a
+    // received message has failed to be processed.
+    AckDeadline time.Duration
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Okay, but it's also a property of SubscriptionConfig in GCP. Can you say more about what's wrong with the way it's laid out here, and what should be done instead?

File: internal/docs/pubsub/design.md [hunk 6]
```diff
@@ -0,0 +1,915 @@
...
+package acmepubsub
+
+import (
+    "context"
+
+    "github.com/google/go-cloud/pubsub"
+)
+
+// TopicOptions contains configuration for Topics.
+type TopicOptions struct {
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Done.

File: internal/docs/pubsub/design.md [hunk 7]
```diff
@@ -0,0 +1,915 @@
...
+func (t *topic) SendBatch(ctx context.Context, []*pubsub.Message) error {
+    // ...
+}
+
+func (t *topic) Close(ctx context.Context) error {
+    // ...
+}
+
+// SubscriptionOptions contains configuration for Subscriptions.
+type SubscriptionOptions struct {
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Done.

File: internal/docs/pubsub/design.md [hunk 8]
```diff
@@ -0,0 +1,915 @@
...
+    close(t.doneChan)
+    return t.driver.Close()
+}
+
+// NewTopic makes a pubsub.Topic from a driver.Topic, a batchSize parameter
+// telling the maximum number of messages to send in a batch, and a sendWait
+// parameter telling the max number of time to wait before sending a non-empty
+// batch. Behind the scenes, NewTopic spins up a goroutine to bundle messages
+// into batches and send them to the server.
+func NewTopic(d driver.Topic, batchSize int, sendWait time.Duration) *Topic {
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: I agree that it's excessive for this doc. Removing it and similar excess in `Receive`.

File: internal/docs/pubsub/design.md [hunk 9]
```diff
@@ -0,0 +1,915 @@
...
+* The subscription callback idea does not appear to be compatible with inverted worker pools.
+
+## Acknowledgements
+In pubsub systems with acknowledgement, messages are kept in a queue associated with the subscription on the server. When a client receives one of these messages, its counterpart on the server is marked as being processed. Once the client finishes processing the message, it sends an acknowledgement (or "ack") to the server and the server removes the message from the subscription queue. There may be a deadline for the acknowledgement, past which the server unmarks the message so that it can be received again for another try at processing.
+
+Redis and ZeroMQ don’t support Ack, but many others do including GCP PubSub, Azure Service Bus, and RabbitMQ. Given the wide support and usefulness, it makes sense to support this in Go Cloud.
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/docs/pubsub/design.md]: Good find. Added a link.

**Comment by ijt** [comment]
Okay, I added `Subscription.Update` with support for `AckDeadline`. PTAL.

**Comment by jba** [comment]
The problem is that `Subscription.Update` is an admin feature.

**Comment by ijt** [comment]
@jba , that's a fair point, but isn't updating `AckDeadline` also an admin feature?

**Comment by vangent** [comment]
If we don't support creating Subscriptions, should we support updating them?

**Comment by ijt** [comment]
@vangent , probably not, at least for now. I'm removing the subscription update method now. We can always add it back in later.

**Comment by Sajmani** [comment]
Nice doc, thank you. A few comments.

First code example: the publishing (Sending) of a message is somewhat obscured by the HTTP server around it. Could you just have an example of "here's how you create a client, and here's how you publish a message on a topic"? I'm not sure what the HTTP handler adds to this picture.

If you do keep the HTTP handler, then shouldn't you be passing the request Context as the context argument to Send, so that the Send is canceled if the HTTP request is canceled?

The text below it says "The call to Send will only return after the message has been sent to the server or its sending has failed." This should probably mention that the Send will fail if the context is canceled.

In code example after "A more realistic subscriber client would process messages in a loop, like this:"
Why not just "return nil" instead of "break Loop"?

All the logic around supporting Ctrl-C seems distracting from helping readers understand the API. I encourage you to simplify your code examples to focus on the API behavior, not making the programs "realistic". It's fine to provide fleshed-out example programs alongside the doc.

**Comment by ijt** [comment]
On Mon, Nov 5, 2018 at 8:02 AM Sameer Ajmani <notifications@github.com>
wrote:

> Nice doc, thank you. A few comments.
>
Thanks. :-)

> First code example: the publishing (Sending) of a message is somewhat
> obscured by the HTTP server around it. Could you just have an example of
> "here's how you create a client, and here's how you publish a message on a
> topic"? I'm not sure what the HTTP handler adds to this picture.
>
 I was trying to show an example of how we expect pubsub Send to be used:
notifying about events. I think it works well as a running example and
isn't too distracting. It also highlights how the batch-oriented API
requires even more error-prone app code to sanely handle request
cancellation.

> If you do keep the HTTP handler, then shouldn't you be passing the request
> Context as the context argument to Send, so that the Send is canceled if
> the HTTP request is canceled?
>
Done.

> The text below it says "The call to Send will only return after the
> message has been sent to the server or its sending has failed." This should
> probably mention that the Send will fail if the context is canceled.
>
Done.

> In code example after "A more realistic subscriber client would process
> messages in a loop, like this:"
> Why not just "return nil" instead of "break Loop"?
>
All the logic around supporting Ctrl-C seems distracting from helping
> readers understand the API. I encourage you to simplify your code examples
> to focus on the API behavior, not making the programs "realistic". It's
> fine to provide fleshed-out example programs alongside the doc.
>
Done.

**Pull request merged**